### PR TITLE
[Auth] - Adding Password based authentication

### PR DESCRIPTION
**Pull request by rbb17**
Adding support for password based authentication to trino-gateway.

Enabled in-memory caching-
Mechanism: If a user is active in last 10 minutes then its authentication will be served via in-memory cache, else it will be routed to datum for verification.

### CODE REVIEW

File: internal/config/config.go [hunk 1]
```diff
@@ -27,8 +27,10 @@ type App struct {
 }
 
 type Auth struct {
-	Token          string
-	TokenHeaderKey string
+	Token               string
+	TokenHeaderKey      string
+	ValidationURL       string
```
Comment (utk-spartan) [inline internal/config/config.go]: both the auth setup are separate 
one is for gatewayserver other for router

track its config separately, either under `Auth.router`

or `router.Auth`

File: internal/router/auth.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+type AuthCache struct {
+	cache map[string]struct {
+		Timestamp time.Time
+		Password  string
+	}
+	mu sync.Mutex
+}
+
+/*
+Calls an authentication api,
```
Comment (utk-spartan) [inline internal/router/auth.go]: Let's call it auth token validator service

File: internal/router/auth.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+				Timestamp time.Time
+				Password  string
+			}),
+		}
+		*ctx = context.WithValue(*ctx, "authCache", authCache)
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		var trinoUser = "X-Trino-User"
+		var trinoPassword = "X-Trino-Password"
```
Comment (utk-spartan) [inline internal/router/auth.go]: move these to `trinoheaders` package as constants

File: internal/router/auth.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+func (authCache *AuthCache) check(username, password string) bool {
+	authCache.mu.Lock()
+	defer authCache.mu.Unlock()
+
+	entry, found := authCache.cache[username]
+	if !found {
+		return false
+	}
+
+	if time.Since(entry.Timestamp) > 10*time.Minute {
```
Comment (utk-spartan) [inline internal/router/auth.go]: the time period needs to be configurable via app config

File: internal/router/auth.go [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		var trinoUser = "X-Trino-User"
+		var trinoPassword = "X-Trino-Password"
+
+		username := r.Header.Get(trinoUser)
+		password := r.Header.Get(trinoPassword)
+
+		if authCache.check(username, password) {
```
Comment (utk-spartan) [inline internal/router/auth.go]: Refactor, move this to `authenticate method

This method `WithAuth` will only extract values from request headers and call the `authenticate` method,
and use the returned values from it

**Review by utk-spartan — CHANGES_REQUESTED** [review]

File: internal/config/config.go (hunk 1 above)
Comment (rbb17) [reply internal/config/config.go]: OK, will change it to `auth.router`

File: internal/router/auth.go (hunk 3 above)
Comment (rbb17) [reply internal/router/auth.go]: ok

File: internal/router/auth.go (hunk 4 above)
Comment (rbb17) [reply internal/router/auth.go]: Ok

File: internal/config/config.go [hunk 6]
```diff
@@ -29,6 +29,11 @@ type App struct {
 type Auth struct {
 	Token          string
 	TokenHeaderKey string
+	Router         struct {
+		ValidationURL   string
+		ValidationToken string
+		CachedInterval  string
```
Comment (utk-spartan) [inline internal/config/config.go]: rename to CacheTTL

File: internal/router/auth.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+If Authenticated - {"ok": true}
+If not authenticated- {"ok": false}
+
+@returns-
+boolean{True or False},error_message
+*/
+func authenticate(ctx *context.Context, username string, password string) (bool, error) {
+	authCache, ok := (*ctx).Value("authCache").(*AuthCache)
+
+	fmt.Println("Ganapati Bappa Moraya", authCache)
```
Comment (utk-spartan) [inline internal/router/auth.go]: remove println

File: internal/router/trinoheaders/trino.go [hunk 8]
```diff
@@ -13,6 +13,8 @@ const (
 	ClientTags           = "Client-Tags"
 	ConnectionProperties = "Connection-Properties"
 	TransactionId        = "Transaction-Id"
+	TrinoUser            = "User"
```
Comment (utk-spartan) [inline internal/router/trinoheaders/trino.go]: drop the Trino prefix

File: internal/router/trinoheaders/trino.go [hunk 9]
```diff
@@ -13,6 +13,8 @@ const (
...
+	TrinoPassword        = "Password"
```
Comment (utk-spartan) [inline internal/router/trinoheaders/trino.go]: drop Trino prefix

File: internal/router/auth.go [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+}
+api returns-
+If Authenticated - {"ok": true}
+If not authenticated- {"ok": false}
+
+@returns-
+boolean{True or False},error_message
+*/
+func authenticate(ctx *context.Context, username string, password string) (bool, error) {
+	authCache, ok := (*ctx).Value("authCache").(*AuthCache)
```
Comment (utk-spartan) [inline internal/router/auth.go]: rename entry in context as `routerAuthCache`

File: internal/config/config.go (hunk 6 above)
Comment (utk-spartan) [reply internal/config/config.go]: suffix time unit 
eg: CacheTTLSeconds or CacheTTLMinutes

**Review by utk-spartan — CHANGES_REQUESTED** [review]

File: internal/router/auth.go (hunk 7 above)
Comment (rbb17) [reply internal/router/auth.go]: 🤦

File: internal/router/trinoheaders/trino.go (hunk 9 above)
Comment (rbb17) [reply internal/router/trinoheaders/trino.go]: 🆗

File: internal/router/auth.go (hunk 7 above)
Comment (rbb17) [reply internal/router/auth.go:43]: 🆗

File: internal/router/auth.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+	"token": "token123"
+}
+api returns-
+If Authenticated - {"ok": true}
+If not authenticated- {"ok": false}
+
+@returns-
+boolean{True or False},error_message
+*/
+func authenticate(ctx *context.Context, username string, password string) (bool, error) {
```
Comment (utk-spartan) [inline internal/router/auth.go]: add unit test

File: internal/router/auth.go [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+	}
+
+	if data.OK {
+		authCache.update(username, password)
+	}
+
+	return data.OK, nil
+}
+
+func (authCache *AuthCache) check(username, password string) bool {
```
Comment (utk-spartan) [inline internal/router/auth.go]: add unit test

File: internal/router/auth.go [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+	if time.Since(entry.Timestamp) > cachedDuration*time.Minute {
+		// If entry is older than cachedDuration, then delete the record and return false
+		delete(authCache.cache, username)
+		return false
+	}
+
+	return entry.Password == password
+}
+
+func (authCache *AuthCache) update(username, password string) {
```
Comment (utk-spartan) [inline internal/router/auth.go]: add unit test

File: internal/config/config.go (hunk 6 above)
Comment (rbb17) [reply internal/config/config.go]: ok , Making it CacheTTLMinutes

File: build/docker/dev/docker-compose.yml [hunk 14]
```diff
@@ -11,6 +11,7 @@ services:
       - ./../../..:/app
     environment:
       APP_ENV: dev-docker
+      TRINO-GATEWAY_AUTH_TOAUTHENTICATE: true
```
Comment (Aka-shi) [inline build/docker/dev/docker-compose.yml]: nit: replace with `TRINO_GATEWAY_AUTH_AUTHENTICATE`

Comment (rbb17) [reply build/docker/dev/docker-compose.yml]: `TRINO-GATEWAY` is code generated prefix for all variables.

Replacing to `TRINO-GATEWAY_AUTH_AUTHENTICATE`

File: internal/router/router.go [hunk 15]
```diff
@@ -108,9 +109,17 @@ func Server(ctx *context.Context, port int, apiClient *GatewayApiClient, routerH
 		},
 	}
 
-	return &http.Server{
-		Handler: &reverseProxy,
+	if boot.Config.Auth.ToAuthenticate {
```
Comment (utk-spartan) [inline internal/router/router.go]: `boot.Config.Auth.ToAuthenticate`
pass it from main.go via argument

File: internal/config/config.go [hunk 16]
```diff
@@ -27,8 +27,14 @@ type App struct {
 }
 
 type Auth struct {
+	Authenticate   bool
```
Comment (utk-spartan) [inline internal/config/config.go]: move it inside `Router`

File: build/docker/dev/Dockerfile [hunk 17]
```diff
@@ -4,7 +4,7 @@ WORKDIR /app
 
 
 RUN apk update \
-    && apk add --no-cache bash make protobuf protobuf-dev git gzip curl
+    && apk add --no-cache bash make protobuf protobuf-dev git gzip curl build-base
```
Comment (utk-spartan) [inline build/docker/dev/Dockerfile:7]: why is build-base needed?

**Review by utk-spartan — CHANGES_REQUESTED** [review]

File: internal/config/config.go (hunk 16 above)
Comment (rbb17) [reply internal/config/config.go]: ack

File: build/docker/dev/Dockerfile (hunk 17 above)
Comment (rbb17) [reply build/docker/dev/Dockerfile:7]: I was unable run  (existing / added)  unit tests also without it, not completely sure why its needed.

File: internal/router/router.go (hunk 15 above)
Comment (rbb17) [reply internal/router/router.go]: got it

**Review by utk-spartan — APPROVED** [review]

**Pull request merged**